[CodeGen] Add ValueType v3i8 (NFCI).

fpetrogalli opened the pull request.

fpetrogalli committed:
[CodeGen] Add ValueType v3i8 (NFCI).

fpetrogalli commented:
Hang on, genuine failures in https://buildkite.com/llvm-project/github-pull-requests/builds/36346#018d7e21-6650-4dd6-b687-9f472dbd9145


```
  LLVM :: CodeGen/AArch64/vec3-loads-ext-trunc-stores.ll
```

fpetrogalli committed:
Update getTypeForEVT to fix failures caused by the missing case for `v3i8`.

fpetrogalli commented:
The current version will fail these two AMDGPU CodeGen tests: 

```
  LLVM :: CodeGen/AMDGPU/calling-conventions.ll
  LLVM :: CodeGen/AMDGPU/store-v3i64.ll
```

@arsenm, are you able to look into this? I blamed places in the AMDGPU backend that mention truncating stores to i8, you seem to have workedd with them. Can you help out here? My knowledge of this ISA is close to zero...

arsenm commented:
You probably need to update the setTruncStoreAction/setLoadExtActions, the current set go through the existing i8 vectors

fpetrogalli committed:
[AMDGPU] Handle truncating store of v3i64 and v3i32 to v3i8.

Fixes the tests:

  LLVM :: CodeGen/AMDGPU/calling-conventions.ll
  LLVM :: CodeGen/AMDGPU/store-v3i64.ll

fpetrogalli commented:
> You probably need to update the setTruncStoreAction/setLoadExtActions, the current set go through the existing i8 vectors

Thank you @arsenm , that did it locally. Let's see what buildkite has to say.

Francesco

fpetrogalli committed:
Merge branch 'main' into add-v3i8-valuetype

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -308,8 +308,11 @@ AMDGPUTargetLowering::AMDGPUTargetLowering(const TargetMachine &TM,
   setTruncStoreAction(MVT::v2f64, MVT::v2f32, Expand);
   setTruncStoreAction(MVT::v2f64, MVT::v2f16, Expand);
 
+  setTruncStoreAction(MVT::v3i32, MVT::v3i8, Expand);
```
I suspect we're missing test coverage (or we're doing unnecessary work) for the load half of this

fpetrogalli replied:
@arsenm - may I merge the patch without taking care of this?

arsenm replied:
yes

fpetrogalli committed:
Merge branch 'main' into add-v3i8-valuetype
